pin/gc/gc.go

```diff
@@ -0,0 +1,102 @@
+package gc
+
+import (
+	bstore "github.com/ipfs/go-ipfs/blocks/blockstore"
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	bserv "github.com/ipfs/go-ipfs/blockservice"
+	offline "github.com/ipfs/go-ipfs/exchange/offline"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	pin "github.com/ipfs/go-ipfs/pin"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+	eventlog "github.com/ipfs/go-ipfs/thirdparty/eventlog"
+)
+
+var log = eventlog.Logger("gc")
+
+// GC performs a mark and sweep garbage collection of the blocks in the blockstore
+// first, it creates a 'marked' set and adds to it the following:
+// - all recursively pinned blocks, plus all of their descendants (recursively)
+// - all directly pinned blocks
+// - all blocks utilized internally by the pinner
+//
+// The routine then iterates over every block in the blockstore and
+// deletes any block that is not found in the marked set.
+func GC(ctx context.Context, bs bstore.GCBlockstore, pn pin.Pinner) (<-chan key.Key, error) {
```

Comment:
actually yeah, so the pin set + these functions + a few cmds is all we need to implement:

- `ipfs provide [-r] <path> - mark path as provided -- publicly announced`
- `ipfs serve [-r] <path> - mark path as served`
- `ipfs serve [-r] <path> <peer-set> - mark path as served to <peer-set>` <-- nice.

we need a word for this, maybe...
- {pin, provide, serve} {recursive, direct, indirect} {label, labeling}
- {pin, provide, serve} {recursive, direct, indirect} {color, coloring}  (from graph theory)
- {pin, provide, serve} {recursive, direct, indirect} {mark, marking}  
- {pin, provide, serve} {recursive, direct, indirect} {tag, tagging}  (though tag has other connotations)
- {pin, provide, serve} {recursive, direct, indirect} {group, grouping}
- {pin, provide, serve} {recursive, direct, indirect} {set}
- {pin, provide, serve} {recursive, direct, indirect} {record}  // since they will be records, either local or networked